File: problems/tier1-foundation/001-payment-ranker/solution.js
// Payment ranker — Strategy pattern reference solution (JavaScript).

class PaymentMethod {
  constructor(name, cashbackRate, transactionFee, usageCount, easyRefundEligible = false) {
    this.name = name;
    this.cashbackRate = cashbackRate;
    this.transactionFee = transactionFee;
    this.usageCount = usageCount;
    this.easyRefundEligible = easyRefundEligible;
  }
}

class RankingStrategy {
  // Return true iff `a` should rank strictly before `b`.
  compare(a, b) { throw new Error('not implemented'); }
}

class RewardsMaximizer extends RankingStrategy {
  compare(a, b) { return a.cashbackRate > b.cashbackRate; }
}

class LowFeeSeeker extends RankingStrategy {
  compare(a, b) { return a.transactionFee < b.transactionFee; }
}

class TrustBasedRanker extends RankingStrategy {
  compare(a, b) { return a.usageCount > b.usageCount; }
}

class CompositeStrategy extends RankingStrategy {
  constructor(criteria) { super(); this.criteria = criteria; }
  compare(a, b) {
    for (const s of this.criteria) {
      if (s.compare(a, b)) return true;
      if (s.compare(b, a)) return false;
    }
    return false;
  }
}

class EasyRefundStrategy extends RankingStrategy {
  constructor(prefer) { super(); this.prefer = prefer; }
  compare(a, b) {
    if (!this.prefer) return false;
    return a.easyRefundEligible && !b.easyRefundEligible;
  }
}

class PaymentRanker {
  constructor(strategy) { this.strategy = strategy; }
  setStrategy(strategy) { this.strategy = strategy; }
  rank(methods) {
    // Translate strict-less comparator into a stable sort via insertion.
    const result = [];
    for (const m of methods) {
      let inserted = false;
      for (let i = 0; i < result.length; i++) {
        if (this.strategy.compare(m, result[i])) {
          result.splice(i, 0, m);
          inserted = true;
          break;
        }
      }
      if (!inserted) result.push(m);
    }
    return result;
  }
}

function rank_by_rewards(methods) {
  return new PaymentRanker(new RewardsMaximizer()).rank(methods);
}

function rank_by_low_fee(methods) {
  return new PaymentRanker(new LowFeeSeeker()).rank(methods);
}

function rank_by_trust(methods) {
  return new PaymentRanker(new TrustBasedRanker()).rank(methods);
}

function rank_composite(methods, criteria) {
  return new PaymentRanker(new CompositeStrategy(criteria)).rank(methods);
}

function rank_with_refund_filter(methods, preferEasyRefund) {
  const composite = new CompositeStrategy([
    new EasyRefundStrategy(preferEasyRefund),
    new RewardsMaximizer(),
  ]);
  return new PaymentRanker(composite).rank(methods);
}

module.exports = {
  PaymentMethod, RankingStrategy, RewardsMaximizer, LowFeeSeeker, TrustBasedRanker,
  CompositeStrategy, EasyRefundStrategy, PaymentRanker,
  rank_by_rewards, rank_by_low_fee, rank_by_trust, rank_composite, rank_with_refund_filter,
};

File: problems/tier1-foundation/001-payment-ranker/solution.py
"""Payment ranker — Strategy pattern reference solution (Python)."""

from abc import ABC, abstractmethod


class PaymentMethod:
    def __init__(self, name, cashbackRate, transactionFee, usageCount,
                 easyRefundEligible=False):
        self.name = name
        self.cashbackRate = cashbackRate
        self.transactionFee = transactionFee
        self.usageCount = usageCount
        self.easyRefundEligible = easyRefundEligible


class RankingStrategy(ABC):
    @abstractmethod
    def compare(self, a, b):
        """Return True iff `a` should rank strictly before `b`."""


class RewardsMaximizer(RankingStrategy):
    def compare(self, a, b):
        return a.cashbackRate > b.cashbackRate


class LowFeeSeeker(RankingStrategy):
    def compare(self, a, b):
        return a.transactionFee < b.transactionFee


class TrustBasedRanker(RankingStrategy):
    def compare(self, a, b):
        return a.usageCount > b.usageCount


class CompositeStrategy(RankingStrategy):
    def __init__(self, criteria):
        self.criteria = criteria

    def compare(self, a, b):
        for s in self.criteria:
            if s.compare(a, b):
                return True
            if s.compare(b, a):
                return False
        return False


class EasyRefundStrategy(RankingStrategy):
    def __init__(self, prefer):
        self.prefer = prefer

    def compare(self, a, b):
        if not self.prefer:
            return False
        return a.easyRefundEligible and not b.easyRefundEligible


class PaymentRanker:
    def __init__(self, strategy):
        self.strategy = strategy

    def set_strategy(self, strategy):
        self.strategy = strategy

    def rank(self, methods):
        # Translate strict-less comparator into a stable sort via insertion.
        items = list(methods)
        result = []
        for m in items:
            inserted = False
            for i, existing in enumerate(result):
                if self.strategy.compare(m, existing):
                    result.insert(i, m)
                    inserted = True
                    break
            if not inserted:
                result.append(m)
        return result


def rank_by_rewards(methods):
    return PaymentRanker(RewardsMaximizer()).rank(methods)


def rank_by_low_fee(methods):
    return PaymentRanker(LowFeeSeeker()).rank(methods)


def rank_by_trust(methods):
    return PaymentRanker(TrustBasedRanker()).rank(methods)


def rank_composite(methods, criteria):
    return PaymentRanker(CompositeStrategy(criteria)).rank(methods)


def rank_with_refund_filter(methods, preferEasyRefund):
    composite = CompositeStrategy([
        EasyRefundStrategy(preferEasyRefund),
        RewardsMaximizer(),
    ])
    return PaymentRanker(composite).rank(methods)
